libraries/AP_HAL_Linux/Poller.cpp

```diff
@@ -0,0 +1,403 @@
+            // fast byte scan instead.  This should happen more often.
+            auto marker_pos = memchr(ptr, mavlink_marker, limit);
+            if (marker_pos) {
+                return static_cast<uint32_t>(
+                    static_cast<const uint8_t *>(marker_pos) - ptr);
+            }
+        } else {
+            for (uint32_t i = 0; i < limit; i++) {
+                if (_write_buffer.peek(i) == mavlink_marker) {
+                    return i;
+                }
+            }
+        }
+
+        // No MAVLink marker, limit the send size to mavlink_max_size.
+        return limit;
+    }
+
+    if (available < mavlink_hdr_size) {
+        return 0; // Not a full MAVLink packet yet.
+    }
+
+    // Possible MAVLink packet, just check if it is complete.
+    const auto pktlen = _write_buffer.peek(1); // Length is on 2nd byte.
+    if (pktlen == -1 || available < static_cast<uint8_t>(pktlen) + mavlink_hdr_size) {
+        return 0; // Not a full MAVLink packet yet.
+    }
+
+    // Packet seems complete. Send one at a time.
+    return pktlen + mavlink_hdr_size;
+}
+
+void PacketedBufferedPollable::on_can_write(int max_time_ms) {
+    if (_write_sem.take(max_time_ms)) {
+        write_fd(to_mavlink_boundary(_write_buffer.available()));
+        _write_sem.give();
+    }
+}
+
+TCPServerPollable::TCPServerPollable(const char *ip, uint16_t port) : Pollable(-1) {
+    _fd = socket(AF_INET, SOCK_STREAM | SOCK_CLOEXEC, 0);
+    if (_fd < 0) {
+        return;
+    }
+
+    int one = 1;
+    setsockopt(_fd, IPPROTO_TCP, TCP_NODELAY, &one, sizeof(one));
+}
+
+bool TCPServerPollable::bind(const char *address, int port) {
+    if (_fd < 0) {
+        return false;
+    }
+
+    sockaddr_in saddrin = {0};
+
+    saddrin.sin_port = htons(port);
+    saddrin.sin_family = AF_INET;
+    saddrin.sin_addr.s_addr = inet_addr(address);
+#ifdef HAVE_SOCK_SIN_LEN
+    saddrin.sin_len = sizeof(sockaddr_in);
+#endif
+
+    return ::bind(_fd, reinterpret_cast<const sockaddr *>(&saddrin), sizeof(saddrin)) == 0;
+}
+
+void TCPServerPollable::reuse_address(bool setting) {
+    if (_fd < 0) {
+        return;
+    }
+
+    int int_setting = setting;
+    setsockopt(_fd, SOL_SOCKET, SO_REUSEADDR, &int_setting, sizeof(int_setting));
+}
+
+bool TCPServerPollable::listen(int backlog_size) {
+    if (_fd < 0) {
+        return false;
+    }
+
+    return ::listen(_fd, backlog_size) == 0;
+}
+
+BufferedPollable *TCPServerPollable::accept(bool blocking)
+{
+    if (_client) {
+        return _client;
+    }
+
+    if (blocking) {
+        int fd = ::accept(_fd, NULL, NULL);
+
+        if (fd >= 0) {
+            int one = 1;
+            setsockopt(fd, IPPROTO_TCP, TCP_NODELAY, &one, sizeof(one));
+
+            return new BufferedPollable(fd);
+        }
+    }
+
+    return nullptr;
+}
+
+void TCPServerPollable::on_can_read(int max_timeout_ms) {
+    delete _client;
+    _client = nullptr;
```

Comment:
couldn't we just invalidate the BufferedPollable instead of freeing + allocating?